src/sagemaker/workflow/steps.py

```diff
@@ -496,6 +496,7 @@ def __init__(
         self.job_arguments = job_arguments
         self.code = code
         self.property_files = property_files
+        self.job_name = name
```

Comment:
If we use the step name as the job name then all input and output locations will be overwritten when the args are normalized. In practical terms that means any pipeline with a step named `MyProcessingStep` will start using the code for this processor.

Things get a little better if we include the `pipeline_name` in the `job_name`, but that still means the code will be overwritten which will prevent recreatibility (ie, looking at a previous execution and going to S3 to see the script that was run).

Let me give this some thought. We need some signal of intentionality from the user here to maintain backwards compatibility.

Reply:
I think the best approach here is to hash the caller's script contents and include that in the `job_name`. This ensures there will be cache hits/misses as appropriate while not requiring any input from the user. I can work on this change since it's a little heavier weight than initially expected

Reply:
Good callout! How do you feel about using the enable caching boolean as the signal of intentionality from the user? Something like:

```
name = pipeline_name + job_name
self.job_name = name if cache_config.enable_caching else None
```

Reply:
Your comment loaded later for me - I like your idea as well. This will go against the documentation [here](https://docs.aws.amazon.com/sagemaker/latest/dg/pipelines-caching.html)  

> Pipelines doesn't check whether the data or code that the arguments point to has changed

As a user though, I do believe it will be nice to evolve caching pipeline steps functionality to detect data/code changes!

Reply:
https://github.com/aws/sagemaker-python-sdk/pull/2790

I'm going to close this PR